plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineFlexFacetTest.java

```diff
@@ -65,4 +69,53 @@ public void testFacetLabel() {
     IProjectFacet projectFacet = ProjectFacetsManager.getProjectFacet(AppEngineFlexFacet.ID);
     Assert.assertEquals("App Engine Java Flexible Environment", projectFacet.getLabel());
   }
+
+  @Test
+  public void testInstallConstraints_okWithJava8Servlet25() {
+    IConstraint constraint = AppEngineFlexFacet.FACET_VERSION.getConstraint();
+    IStatus result = constraint.check(Arrays.asList(JavaFacet.VERSION_1_8, WebFacetUtils.WEB_25));
+    Assert.assertTrue(result.isOK());
+  }
+
+  @Test
+  public void testInstallConstraints_okWithJava8Servlet30() {
+    IConstraint constraint = AppEngineFlexFacet.FACET_VERSION.getConstraint();
+    IStatus result = constraint.check(Arrays.asList(JavaFacet.VERSION_1_8, WebFacetUtils.WEB_30));
+    Assert.assertTrue(result.isOK());
+  }
+
+  @Test
+  public void testInstallConstraints_okWithJava8Servlet31() {
+    IConstraint constraint = AppEngineFlexFacet.FACET_VERSION.getConstraint();
+    IStatus result = constraint.check(Arrays.asList(JavaFacet.VERSION_1_8, WebFacetUtils.WEB_31));
+    Assert.assertTrue(result.isOK());
+  }
+
+  @Test
+  public void testInstallConstraints_notOkWithNoJava() {
+    IConstraint constraint = AppEngineFlexFacet.FACET_VERSION.getConstraint();
+    IStatus result = constraint.check(Arrays.asList(WebFacetUtils.WEB_31));
+    Assert.assertFalse(result.isOK());
+  }
+
+  @Test
+  public void testInstallConstraints_notOkWithNoServlet() {
+    IConstraint constraint = AppEngineFlexFacet.FACET_VERSION.getConstraint();
+    IStatus result = constraint.check(Arrays.asList(JavaFacet.VERSION_1_8));
+    Assert.assertFalse(result.isOK());
+  }
+
+  @Test
+  public void testInstallConstraints_notOkWithJava7Servlet31() {
+    IConstraint constraint = AppEngineFlexFacet.FACET_VERSION.getConstraint();
+    IStatus result = constraint.check(Arrays.asList(JavaFacet.VERSION_1_7, WebFacetUtils.WEB_31));
```

Comment:
My understanding is that we don't have a Java 7 runtime: https://cloud.google.com/appengine/docs/flexible/java/dev-jetty9. The Java 8 / Jetty 9.3 Runtime for WARs, and Java 8 Runtime for JARs.